### ISSUE TITLE
What's the difference between val.py and detect.py?

### ISSUE THREAD

**Issue by Henry0528** [ultralytics/yolov5#8584]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

For the same image, I got different results from val.py and detect.py.
I use the same model and the same model weights
the result from val.py
<img width="695" alt="image" src="https://user-images.githubusercontent.com/61041040/179173608-496455fd-54c4-4f50-aadd-cebe051b08bf.png">
the result from detect.py
<img width="1087" alt="image" src="https://user-images.githubusercontent.com/61041040/179173952-1c41f772-32b7-4ba5-814e-9a2ca6466f47.png">

**Comment by glenn-jocher** [comment]
@Henry0528 👋 Hello, thanks for asking about the differences between [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py), [detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py) and [val.py](https://github.com/ultralytics/yolov5/blob/master/val.py) in YOLOv5 🚀.

These 3 files are designed for different purposes and utilize different dataloaders with different settings. train.py dataloaders are designed for a speed-accuracy compromise, val.py is designed to obtain the best mAP on a validation dataset, and detect.py is designed for best real-world inference results. A few important aspects of each:


## train.py

- **trainloader** LoadImagesAndLabels(): designed to load train dataset images and labels. Augmentation capable and enabled. https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/train.py#L210-L213 **val_loader** LoadImagesAndLabels(): designed to load val dataset images and labels. Augmentation capable but disabled. https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/train.py#L220-L223
- image size: `640`
- rectangular inference: `False`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): `0.001`
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): `0.6`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): `True`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): `None`

## val.py

- **dataloader** LoadImagesAndLabels(): designed to load train, val, test dataset images and labels. Augmentation capable but disabled. 
- https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/val.py#L152-L153
- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): `0.001`
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): `0.6`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): `True`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): `0.5 * maximum stride`


## detect.py

- **dataloaders** (multiple): designed for loading multiple types of media (images, videos, globs, directories, streams). https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/detect.py#L120-L128
- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L152): `0.25` 
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L153): `0.45`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/utils/general.py#L415): `False`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L45-L54): `None`


## YOLOv5 PyTorch Hub Inference

YOLOv5 PyTorch Hub models are`AutoShape()` instances used for image loading, preprocessing, inference and NMS. For more info see [YOLOv5 PyTorch Hub Tutorial](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading)
https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/models/common.py#L276-L282

- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `0.25 `
- [iou threshold](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `0.45`
- [multi-label](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L288): `False`
- [padding](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `None`

Good luck 🍀 and let us know if you have any other questions!

**Comment by frabob2017** [comment]
This explains the result  difference between val.py and model = torch.hub().  Which one is more sensitive for detection, val.py or detection.py?

**Comment by glenn-jocher** [comment]
@frabob2017 both `val.py` and `detect.py` are sensitive for detection, but they are optimized for different use cases. 

`val.py` is designed for computing the mAP (mean average precision) metric on validation datasets and is therefore optimized for accuracy and precision. It also performs rectangular inference.

`detect.py`, on the other hand, is designed for real-world inference and is optimized for fast, high-performance object detection in various scenarios. It has different default confidence and IoU thresholds as well as other detection settings to optimize for real-world performance.

If you are looking for the most accurate detection results for validation and testing datasets, `val.py` might be more suitable. However, for real-world inference scenarios, `detect.py` may be more suitable due to its optimized performance settings.

Hope this helps clarify things! Let me know if you have further questions.
